golf2/src/aes.rs

```diff
@@ -0,0 +1,269 @@
+                    .crypt(input_buffer.as_ref())
+                    .map_err(|_| SyscallError::InternalError));
+
+                Ok(size as isize)
+            })
+            .unwrap_or(Err(SyscallError::InternalError))
+    }
+
+    fn read_data(&self, caller_id: AppId) -> Result<isize, SyscallError> {
+        self.apps
+            .enter(caller_id, |app_data, _| {
+                match self.current_user.get() {
+                    Some(cur) if cur.idx() == caller_id.idx() => {}
+                    _ => return Err(SyscallError::InvalidState),
+                }
+
+                let output_buffer = match app_data.output_buffer {
+                    Some(ref mut slice) => slice,
+                    None => return Err(SyscallError::InvalidState),
+                };
+
+                let size = try!(self.device
+                    .read_data(output_buffer.as_mut())
+                    .map_err(|_| SyscallError::InternalError));
+
+                Ok(size as isize)
+            })
+            .unwrap_or(Err(SyscallError::InternalError))
+    }
+
+    fn finish(&self, caller_id: AppId) -> Result<isize, SyscallError> {
+        self.apps
+            .enter(caller_id, |_, _| {
+                match self.current_user.get() {
+                    Some(cur) if cur.idx() == caller_id.idx() => {}
+                    _ => return Err(SyscallError::InvalidState),
+                }
+
+                self.current_user.set(None);
+
+                try!(self.device.finish().map_err(|_| SyscallError::InternalError));
+
+                Ok(0)
+            })
+            .unwrap_or(Err(SyscallError::InternalError))
+    }
+
+    fn register(&self, interrupt: Interrupt, callback: Callback) -> isize {
+        let _ = self.apps.enter(callback.app_id(), |app_data, _| {
+            let ref mut cb = app_data.callbacks;
+            match interrupt {
+                Interrupt::WFIFOOverflow => cb.wfifo_overflow = Some(callback),
+                Interrupt::RFIFOOverflow => cb.rfifo_overflow = Some(callback),
+                Interrupt::RFIFOUnderflow => cb.rfifo_underflow = Some(callback),
+                Interrupt::DoneCipher => cb.done_cipher = Some(callback),
+                Interrupt::DoneKeyExpansion => cb.done_key_expansion = Some(callback),
+                Interrupt::DoneWipeSecrets => cb.done_wipe_secrets = Some(callback),
+            }
+        });
+
+        0
+    }
+}
+
+impl<'a> Driver for AesDriver<'a> {
+    fn subscribe(&self, subscribe_num: usize, callback: Callback) -> isize {
+        match subscribe_num {
+            0 => self.register(Interrupt::WFIFOOverflow, callback),
+            1 => self.register(Interrupt::RFIFOOverflow, callback),
+            2 => self.register(Interrupt::RFIFOUnderflow, callback),
+            3 => self.register(Interrupt::DoneCipher, callback),
+            4 => self.register(Interrupt::DoneKeyExpansion, callback),
+            5 => self.register(Interrupt::DoneWipeSecrets, callback),
```

Comment:
Many of these events don't seem to have any particular relevance to userland (what it's going to do or care about the FIFOs overflowing?), so I'm not sure if they should even be exposed, unless there's a need.